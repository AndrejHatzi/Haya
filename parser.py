from sly import Lexer
from sly import Parser

class BasicLexer(Lexer):
    tokens = { NAME, NUMBER, FLOAT, STRING, IF, THEN, ELSE, FOR, FASTERLOOP, WHILE, CREATE, LEQ, PRINT, FUN,TO, ARROW, EQEQ, RPAREN, LPAREN}
    #{ NAME, NUMBER, FLOAT, STRING, IF, THEN, ELSE, FOR, FASTERLOOP, WHILE,FUN,CREATE,TO, ARROW,PRINT,EQEQ, RPAREN, LPAREN}
    ignore = '\t '
    # ADDED '.' TRY FOR FLOATS
    literals = { '=', '+', '-', '/', '*', '(', ')', ',', ';' , '.'}

    # Define tokens
    IF = r'if'
    THEN = r':'
    ELSE = r'else'
    FOR = r'for'
    FASTERLOOP = r'fasterfor'
    WHILE = r'while'
    FUN = r'FUN'
    CREATE = r'createfile'
    TO = r','
    ARROW = r'->'
    PRINT = r'print'
    NAME = r'[a-zA-Z_][a-zA-Z0-9_]*'
    STRING = r'\".*?\"'
    RPAREN = r'[(]'
    LPAREN = r'[)]'
    '''
    @_('expr PLUS term',
   'expr MINUS term')
def expr(self, p):
    return (p[1], p.expr, p.term)
    '''

    '''
    @_('LPAREN expr RPAREN')
def expr(self, p):
    return ('group-expression',p.expr])
    '''


    EQEQ = r'=='
    LEQ = r'<='

    @_(r'\d+')
    def NUMBER(self, t):
        t.value = int(t.value)
        return t

    #match floats
    #Not working!! #=> implementar build 0
    @_(r'[+-]((\d+(\.\d*)?)|\.\d+)([eE][+-]?[0-9]+)?')
    def FLOAT(self, t):
        t.value = float(t.value)
        return t

    @_(r'#.*')
    def COMMENT(self, t):
        pass

    @_(r'\n+')
    def newline(self,t ):
        self.lineno = t.value.count('\n')

class BasicParser(Parser):
    tokens = BasicLexer.tokens

    precedence = (
        ('left', '+', '-'),
        ('left', '*', '/'),
        ('right', 'UMINUS'),
        )

    def __init__(self):
        self.env = { }

    @_('')
    def statement(self, p):
        pass

    @_('PRINT RPAREN NAME LPAREN')
    def statement(self, p):
        return ('print_stmt_var', p.NAME)

    @_('PRINT RPAREN expr LPAREN')
    def statement(self, p):
        return ('print_stmt_expr', p.expr)

    @_('PRINT RPAREN STRING LPAREN')
    def statement(self, p):
        return ('print_stmt_string', p.STRING)

    @_('CREATE STRING')
    def statement(self, p):
        return ('create_stmt_string', p.STRING)

    @_('CREATE RPAREN NAME LPAREN')
    def statement(self, p):
        return ('create_stmt_var', p.NAME)


    #for (x=5, 10): x
    @_('FOR RPAREN var_assign TO expr LPAREN THEN statement')
    def statement(self, p):
        return ('for_loop', ('for_loop_setup', p.var_assign, p.expr), p.statement)

    @_('FASTERLOOP RPAREN var_assign TO expr LPAREN THEN statement')
    def statement(self, p):
        return ('faster_for_loop', ('for_loop_setup', p.var_assign, p.expr), p.statement)

    @_('WHILE RPAREN condition LPAREN THEN statement')
    def statement(self, p):
        return ('while_loop', ('while_loop_setup', p.condition), p.statement)
    #if x==5: x
    @_('IF condition THEN statement')
    def statement(self, p):
        return ('if_stmt', p.condition, ('branch', p.statement))

    #if x==5: x else y
    @_('IF condition THEN statement ELSE statement')
    def statement(self, p):
        return ('if_else_stmt', p.condition, ('branch', p.statement0, p.statement1))

    #Deprecated
    @_('FUN NAME RPAREN LPAREN ARROW statement')
    def statement(self, p):
        return ('fun_def', p.NAME, p.statement)
    #Deprecated
    @_('NAME RPAREN LPAREN')
    def statement(self, p):
        return ('fun_call', p.NAME)

    @_('expr EQEQ expr')
    def condition(self, p):
        return ('condition_eqeq', p.expr0, p.expr1)

    @_('expr LEQ expr')
    def condition(self, p):
        return ('condition_leq', p.expr0, p.expr1)

    @_('var_assign')
    def statement(self, p):
        return p.var_assign

    @_('NAME "=" expr')
    def var_assign(self, p):
        return ('var_assign', p.NAME, p.expr)

    @_('NAME "=" STRING')
    def var_assign(self, p):
        return ('var_assign', p.NAME, p.STRING)

    @_('expr')
    def statement(self, p):
        return (p.expr)

    @_('expr "+" expr')
    def expr(self, p):
        return ('add', p.expr0, p.expr1)

    @_('expr "-" expr')
    def expr(self, p):
        return ('sub', p.expr0, p.expr1)

    @_('expr "*" expr')
    def expr(self, p):
        return ('mul', p.expr0, p.expr1)

    @_('expr "/" expr')
    def expr(self, p):
        return ('div', p.expr0, p.expr1)

    @_('"-" expr %prec UMINUS')
    def expr(self, p):
        return p.expr

    @_('NAME')
    def expr(self, p):
        return ('var', p.NAME)

    @_('NUMBER')
    def expr(self, p):
        return ('num', p.NUMBER)

    @_('FLOAT')
    def expr(self, p):
        return ('float', p.FLOAT)

if __name__ == '__main__':
    lexer = BasicLexer()
    parser = BasicParser()
    env = {}
    while True:
        try:
            text = input('haya console >')
        except EOFError:
            break
        if text:
            tree = parser.parse(lexer.tokenize(text))
            print(tree)
