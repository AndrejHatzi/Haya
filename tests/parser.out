Grammar:

Rule 0     S' -> statement
Rule 1     statement -> expr
Rule 2     statement -> var_assign
Rule 3     statement -> NAME ( )
Rule 4     statement -> FUN NAME ( ) : statement
Rule 5     statement -> <empty>
Rule 6     var_assign -> NAME = statement
Rule 7     var_assign -> NAME = STRING
Rule 8     expr -> NUMBER
Rule 9     expr -> NAME
Rule 10    expr -> - expr  [precedence=right, level=3]
Rule 11    expr -> expr / expr  [precedence=left, level=2]
Rule 12    expr -> expr * expr  [precedence=left, level=2]
Rule 13    expr -> expr - expr  [precedence=left, level=1]
Rule 14    expr -> expr + expr  [precedence=left, level=1]

Terminals, with rules where they appear:

(                    : 3 4
)                    : 3 4
*                    : 12
+                    : 14
-                    : 10 13
/                    : 11
:                    : 4
=                    : 6 7
FUN                  : 4
NAME                 : 3 4 6 7 9
NUMBER               : 8
STRING               : 7
error                : 

Nonterminals, with rules where they appear:

expr                 : 1 10 11 11 12 12 13 13 14 14
statement            : 4 6 0
var_assign           : 2


state 0

    (0) S' -> . statement
    (1) statement -> . expr
    (2) statement -> . var_assign
    (3) statement -> . NAME ( )
    (4) statement -> . FUN NAME ( ) : statement
    (5) statement -> .
    (8) expr -> . NUMBER
    (9) expr -> . NAME
    (10) expr -> . - expr
    (11) expr -> . expr / expr
    (12) expr -> . expr * expr
    (13) expr -> . expr - expr
    (14) expr -> . expr + expr
    (6) var_assign -> . NAME = statement
    (7) var_assign -> . NAME = STRING
    NAME            shift and go to state 4
    FUN             shift and go to state 5
    $end            reduce using rule 5 (statement -> .)
    NUMBER          shift and go to state 6
    -               shift and go to state 7

    statement                      shift and go to state 1
    expr                           shift and go to state 2
    var_assign                     shift and go to state 3

state 1

    (0) S' -> statement .


state 2

    (1) statement -> expr .
    (11) expr -> expr . / expr
    (12) expr -> expr . * expr
    (13) expr -> expr . - expr
    (14) expr -> expr . + expr
    $end            reduce using rule 1 (statement -> expr .)
    /               shift and go to state 8
    *               shift and go to state 9
    -               shift and go to state 10
    +               shift and go to state 11


state 3

    (2) statement -> var_assign .
    $end            reduce using rule 2 (statement -> var_assign .)


state 4

    (3) statement -> NAME . ( )
    (9) expr -> NAME .
    (6) var_assign -> NAME . = statement
    (7) var_assign -> NAME . = STRING
    (               shift and go to state 12
    /               reduce using rule 9 (expr -> NAME .)
    *               reduce using rule 9 (expr -> NAME .)
    -               reduce using rule 9 (expr -> NAME .)
    +               reduce using rule 9 (expr -> NAME .)
    $end            reduce using rule 9 (expr -> NAME .)
    =               shift and go to state 13


state 5

    (4) statement -> FUN . NAME ( ) : statement
    NAME            shift and go to state 14


state 6

    (8) expr -> NUMBER .
    /               reduce using rule 8 (expr -> NUMBER .)
    *               reduce using rule 8 (expr -> NUMBER .)
    -               reduce using rule 8 (expr -> NUMBER .)
    +               reduce using rule 8 (expr -> NUMBER .)
    $end            reduce using rule 8 (expr -> NUMBER .)


state 7

    (10) expr -> - . expr
    (8) expr -> . NUMBER
    (9) expr -> . NAME
    (10) expr -> . - expr
    (11) expr -> . expr / expr
    (12) expr -> . expr * expr
    (13) expr -> . expr - expr
    (14) expr -> . expr + expr
    NUMBER          shift and go to state 6
    NAME            shift and go to state 16
    -               shift and go to state 7

    expr                           shift and go to state 15

state 8

    (11) expr -> expr / . expr
    (8) expr -> . NUMBER
    (9) expr -> . NAME
    (10) expr -> . - expr
    (11) expr -> . expr / expr
    (12) expr -> . expr * expr
    (13) expr -> . expr - expr
    (14) expr -> . expr + expr
    NUMBER          shift and go to state 6
    NAME            shift and go to state 16
    -               shift and go to state 7

    expr                           shift and go to state 17

state 9

    (12) expr -> expr * . expr
    (8) expr -> . NUMBER
    (9) expr -> . NAME
    (10) expr -> . - expr
    (11) expr -> . expr / expr
    (12) expr -> . expr * expr
    (13) expr -> . expr - expr
    (14) expr -> . expr + expr
    NUMBER          shift and go to state 6
    NAME            shift and go to state 16
    -               shift and go to state 7

    expr                           shift and go to state 18

state 10

    (13) expr -> expr - . expr
    (8) expr -> . NUMBER
    (9) expr -> . NAME
    (10) expr -> . - expr
    (11) expr -> . expr / expr
    (12) expr -> . expr * expr
    (13) expr -> . expr - expr
    (14) expr -> . expr + expr
    NUMBER          shift and go to state 6
    NAME            shift and go to state 16
    -               shift and go to state 7

    expr                           shift and go to state 19

state 11

    (14) expr -> expr + . expr
    (8) expr -> . NUMBER
    (9) expr -> . NAME
    (10) expr -> . - expr
    (11) expr -> . expr / expr
    (12) expr -> . expr * expr
    (13) expr -> . expr - expr
    (14) expr -> . expr + expr
    NUMBER          shift and go to state 6
    NAME            shift and go to state 16
    -               shift and go to state 7

    expr                           shift and go to state 20

state 12

    (3) statement -> NAME ( . )
    )               shift and go to state 21


state 13

    (6) var_assign -> NAME = . statement
    (7) var_assign -> NAME = . STRING
    (1) statement -> . expr
    (2) statement -> . var_assign
    (3) statement -> . NAME ( )
    (4) statement -> . FUN NAME ( ) : statement
    (5) statement -> .
    (8) expr -> . NUMBER
    (9) expr -> . NAME
    (10) expr -> . - expr
    (11) expr -> . expr / expr
    (12) expr -> . expr * expr
    (13) expr -> . expr - expr
    (14) expr -> . expr + expr
    (6) var_assign -> . NAME = statement
    (7) var_assign -> . NAME = STRING
    STRING          shift and go to state 23
    NAME            shift and go to state 4
    FUN             shift and go to state 5
    $end            reduce using rule 5 (statement -> .)
    NUMBER          shift and go to state 6
    -               shift and go to state 7

    statement                      shift and go to state 22
    expr                           shift and go to state 2
    var_assign                     shift and go to state 3

state 14

    (4) statement -> FUN NAME . ( ) : statement
    (               shift and go to state 24


state 15

    (10) expr -> - expr .
    (11) expr -> expr . / expr
    (12) expr -> expr . * expr
    (13) expr -> expr . - expr
    (14) expr -> expr . + expr
    /               reduce using rule 10 (expr -> - expr .)
    *               reduce using rule 10 (expr -> - expr .)
    -               reduce using rule 10 (expr -> - expr .)
    +               reduce using rule 10 (expr -> - expr .)
    $end            reduce using rule 10 (expr -> - expr .)


state 16

    (9) expr -> NAME .
    /               reduce using rule 9 (expr -> NAME .)
    *               reduce using rule 9 (expr -> NAME .)
    -               reduce using rule 9 (expr -> NAME .)
    +               reduce using rule 9 (expr -> NAME .)
    $end            reduce using rule 9 (expr -> NAME .)


state 17

    (11) expr -> expr / expr .
    (11) expr -> expr . / expr
    (12) expr -> expr . * expr
    (13) expr -> expr . - expr
    (14) expr -> expr . + expr
    /               reduce using rule 11 (expr -> expr / expr .)
    *               reduce using rule 11 (expr -> expr / expr .)
    -               reduce using rule 11 (expr -> expr / expr .)
    +               reduce using rule 11 (expr -> expr / expr .)
    $end            reduce using rule 11 (expr -> expr / expr .)


state 18

    (12) expr -> expr * expr .
    (11) expr -> expr . / expr
    (12) expr -> expr . * expr
    (13) expr -> expr . - expr
    (14) expr -> expr . + expr
    /               reduce using rule 12 (expr -> expr * expr .)
    *               reduce using rule 12 (expr -> expr * expr .)
    -               reduce using rule 12 (expr -> expr * expr .)
    +               reduce using rule 12 (expr -> expr * expr .)
    $end            reduce using rule 12 (expr -> expr * expr .)


state 19

    (13) expr -> expr - expr .
    (11) expr -> expr . / expr
    (12) expr -> expr . * expr
    (13) expr -> expr . - expr
    (14) expr -> expr . + expr
    -               reduce using rule 13 (expr -> expr - expr .)
    +               reduce using rule 13 (expr -> expr - expr .)
    $end            reduce using rule 13 (expr -> expr - expr .)
    /               shift and go to state 8
    *               shift and go to state 9


state 20

    (14) expr -> expr + expr .
    (11) expr -> expr . / expr
    (12) expr -> expr . * expr
    (13) expr -> expr . - expr
    (14) expr -> expr . + expr
    -               reduce using rule 14 (expr -> expr + expr .)
    +               reduce using rule 14 (expr -> expr + expr .)
    $end            reduce using rule 14 (expr -> expr + expr .)
    /               shift and go to state 8
    *               shift and go to state 9


state 21

    (3) statement -> NAME ( ) .
    $end            reduce using rule 3 (statement -> NAME ( ) .)


state 22

    (6) var_assign -> NAME = statement .
    $end            reduce using rule 6 (var_assign -> NAME = statement .)


state 23

    (7) var_assign -> NAME = STRING .
    $end            reduce using rule 7 (var_assign -> NAME = STRING .)


state 24

    (4) statement -> FUN NAME ( . ) : statement
    )               shift and go to state 25


state 25

    (4) statement -> FUN NAME ( ) . : statement
    :               shift and go to state 26


state 26

    (4) statement -> FUN NAME ( ) : . statement
    (1) statement -> . expr
    (2) statement -> . var_assign
    (3) statement -> . NAME ( )
    (4) statement -> . FUN NAME ( ) : statement
    (5) statement -> .
    (8) expr -> . NUMBER
    (9) expr -> . NAME
    (10) expr -> . - expr
    (11) expr -> . expr / expr
    (12) expr -> . expr * expr
    (13) expr -> . expr - expr
    (14) expr -> . expr + expr
    (6) var_assign -> . NAME = statement
    (7) var_assign -> . NAME = STRING
    NAME            shift and go to state 4
    FUN             shift and go to state 5
    $end            reduce using rule 5 (statement -> .)
    NUMBER          shift and go to state 6
    -               shift and go to state 7

    statement                      shift and go to state 27
    expr                           shift and go to state 2
    var_assign                     shift and go to state 3

state 27

    (4) statement -> FUN NAME ( ) : statement .
    $end            reduce using rule 4 (statement -> FUN NAME ( ) : statement .)
